packages/cli/src/runners/task-managers/task-manager.ts

```diff
@@ -0,0 +1,373 @@
+}
+
+export interface Task {
+	taskId: string;
+	settings: unknown;
+	data: TaskData;
+}
+
+export class JobError extends ApplicationError {}
+
+interface ExecuteFunctionObject {
+	[name: string]: ((...args: unknown[]) => unknown) | ExecuteFunctionObject;
+}
+
+const workflowToParameters = (workflow: Workflow): Omit<WorkflowParameters, 'nodeTypes'> => {
+	return {
+		id: workflow.id,
+		name: workflow.name,
+		active: workflow.active,
+		connections: workflow.connectionsBySourceNode,
+		nodes: Object.values(workflow.nodes),
+		pinData: workflow.pinData,
+		settings: workflow.settings,
+		staticData: workflow.staticData,
+	};
+};
+
+export class TaskManager {
+	requestAcceptRejects: Record<string, { accept: RequestAccept; reject: RequestReject }> = {};
+
+	taskAcceptRejects: Record<string, { accept: TaskAccept; reject: TaskReject }> = {};
+
+	pendingRequests: Record<string, TaskRequest> = {};
+
+	tasks: Record<string, Task> = {};
+
+	async startTask<T>(
+		taskType: string,
+		settings: unknown,
+		executeFunctions: IExecuteFunctions,
+		inputData: ITaskDataConnections,
+		node: INode,
+		workflow: Workflow,
+		runExecutionData: IRunExecutionData,
+		runIndex: number,
+		itemIndex: number,
+		activeNodeName: string,
+		connectionInputData: INodeExecutionData[],
+		siblingParameters: INodeParameters,
+		mode: WorkflowExecuteMode,
+		executeData?: IExecuteData,
+		defaultReturnRunIndex = -1,
+		selfData: IDataObject = {},
+		contextNodeName: string = activeNodeName,
+	): Promise<T> {
+		const data: TaskData = {
+			workflow,
+			runExecutionData,
+			runIndex,
+			connectionInputData,
+			inputData,
+			node,
+			executeFunctions,
+			itemIndex,
+			siblingParameters,
+			mode,
+			executeData,
+			defaultReturnRunIndex,
+			selfData,
+			contextNodeName,
+			activeNodeName,
+		};
+
+		const request: TaskRequest = {
+			requestId: nanoid(),
+			taskType,
+			settings,
+			data,
+		};
+
+		this.pendingRequests[request.requestId] = request;
+
+		const taskIdPromise = new Promise<string>((resolve, reject) => {
+			this.requestAcceptRejects[request.requestId] = {
+				accept: resolve,
+				reject,
+			};
+		});
+
+		this.sendMessage({
+			type: 'requester:taskrequest',
+			requestId: request.requestId,
+			taskType,
+		});
+
+		const taskId = await taskIdPromise;
+
+		const task: Task = {
+			taskId,
+			data,
+			settings,
+		};
+		this.tasks[task.taskId] = task;
+
+		try {
+			const dataPromise = new Promise<TaskResultData>((resolve, reject) => {
+				this.taskAcceptRejects[task.taskId] = {
+					accept: resolve,
+					reject,
+				};
+			});
+
+			this.sendMessage({
+				type: 'requester:tasksettings',
+				taskId,
+				settings,
+			});
+
+			const resultData = await dataPromise;
+			if (resultData.customData) {
+				Object.entries(resultData.customData).forEach(([k, v]) => {
+					if (!runExecutionData.resultData.metadata) {
+						runExecutionData.resultData.metadata = {};
+					}
+					runExecutionData.resultData.metadata[k] = v;
+				});
+			}
```

Comment:
What is this `customData` stuff? Could we have a comment explaining it

Reply:
I've added a comment now, let me know if it's not clear enough.